### PR TITLE
feat: Adding higher_order_moment, batched_outer to Tensorization Functional API

### LINKED ISSUE #22091: higher_order_moment

### LINKED ISSUE #22732: batched_outer

### PR DESCRIPTION
**Pull request by Madjid-CH**
Close #22091
Close #22732

### CODE REVIEW

**Comment by Kibet-g** [comment]
Sure, I can help you with that! To add a higher-order moment calculation to a Tensorization Functional API in Python, you can follow these steps:

Import the necessary libraries:
python
Copy code
import tensorflow as tf
from tensorflow.keras.layers.experimental.preprocessing import TextVectorization
Define a custom layer for calculating the higher-order moment:
python
Copy code
class HigherOrderMomentLayer(tf.keras.layers.Layer):
    def call(self, inputs):
        mean = tf.reduce_mean(inputs, axis=1, keepdims=True)
        centered = inputs - mean
        higher_order_moment = tf.reduce_mean(tf.pow(centered, 3), axis=1)
        return higher_order_moment
Create the Tensorization Functional API with the custom layer:
python
Copy code
input_layer = tf.keras.layers.Input(shape=(None,), dtype=tf.string)
vectorization_layer = TextVectorization(output_mode='tf-idf')(input_layer)
higher_order_moment_layer = HigherOrderMomentLayer()(vectorization_layer)
model = tf.keras.models.Model(inputs=input_layer, outputs=higher_order_moment_layer)
Compile and train your model as needed:
python
Copy code
model.compile(optimizer='adam', loss='mean_squared_error')
model.fit(x_train, y_train, epochs=10, batch_size=32)
Remember to adjust the layers, preprocessing, and model configurations to your specific use case. This example demonstrates adding a higher-order moment calculation using a custom layer in the context of text vectorization, but you can adapt it to other types of data as well.

File: ivy/functional/ivy/experimental/linear_algebra.py [hunk 1]
```diff
@@ -1666,3 +1666,86 @@ def dot(
...
+            current_shape = ivy.shape(tensor)
+            current_size = len(current_shape) - 1
+
+            n_samples = current_shape[0]
+
+            if n_samples != result_shape[0]:
+                raise ValueError(
+                    f"Tensor {i} has a batch-size of {n_samples} but those before had a"
+                    f" batch-size of {result_shape[0]}, all tensors should have the"
+                    " same batch-size."
```
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/linear_algebra.py:1738]: I think I missed this function while creating the ToDo list but I believe this is a general function and should be added to the functional API of Ivy. I have added it to the ToDo list now (https://github.com/unifyai/ivy/issues/21914). Do you mind working on this function as well? You can add both functions to this PR itself.

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Hey @Madjid-CH ! Thanks for working on this issue, this is already a great PR. Just left some minor comments for now. Would you please address these and make the changes to all the places applicable? Feel free to request another review when you are done. Happy coding :)

**Comment by Madjid-CH** [comment]
ivy-gardener

File: ivy/data_classes/array/experimental/linear_algebra.py [hunk 2]
```diff
@@ -675,3 +675,91 @@ def tucker(
...
+        >>> result = ivy.higher_order_moment(a, 3)
+        >>> print(result)
+        ivy.array([[
+            [14, 19],
+            [19, 26]],
+           [[19, 26],
+            [26, 36]
+        ]])
+        """
+        return ivy.higher_order_moment(self._data, order)
```
Comment (hello-fri-end) [inline ivy/data_classes/array/experimental/linear_algebra.py:718]:
```suggestion
        return ivy.higher_order_moment(self._data, order, out=out)
```

File: ivy/functional/ivy/experimental/linear_algebra.py [hunk 3]
```diff
@@ -1666,3 +1666,136 @@ def dot(
...
+        [19, 26]],
+       [[19, 26],
+        [26, 36]
+    ]])
+    """
+    moment = ivy.copy_array(x)
+    for _ in range(order - 1):
+        moment = ivy.batched_outer([moment, x])
+
+    return ivy.mean(moment, axis=0)
```
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/linear_algebra.py:1722]:
```suggestion
    return ivy.mean(moment, axis=0, out=out)
```

File: ivy/functional/ivy/experimental/linear_algebra.py [hunk 4]
```diff
@@ -1666,3 +1666,136 @@ def dot(
...
+            shape_1 = result_shape + (1,) * current_size
+            shape_2 = (n_samples,) + (1,) * result_size + current_shape[1:]
+
+            result = ivy.reshape(result, shape_1) * ivy.reshape(tensor, shape_2)
+        else:
+            result = tensor
+
+        result_shape = ivy.shape(result)
+        result_size = len(result_shape) - 1
+
```
Comment (hello-fri-end) [inline ivy/functional/ivy/experimental/linear_algebra.py:1800]:
```suggestion
if ivy.exists(out):
result = ivy.inplace_update(out, result)
```

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 5]
```diff
@@ -1610,3 +1610,90 @@ def test_tucker_tensorly(tol_norm_2, tol_max_abs, shape, ranks):
...
+            safety_factor_scale="log",
+        )
+    )
+    return dtype, x[0], order
+
+
+@handle_test(
+    fn_tree="functional.ivy.experimental.higher_order_moment",
+    data=_higher_order_moment_data(),
+    test_with_out=st.just(False),
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1635]: After adding the `out` argument in the implementation as I suggested, this wouldn't be required 😄

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 6]
```diff
@@ -1610,3 +1610,90 @@ def test_tucker_tensorly(tol_norm_2, tol_max_abs, shape, ranks):
...
+    input_dtypes, x, order = data
+    helpers.test_function(
+        backend_to_test=backend_fw,
+        test_flags=test_flags,
+        fn_name=fn_name,
+        on_device=on_device,
+        input_dtypes=input_dtypes,
+        x=x,
+        order=order,
+        test_values=False,
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1647]: Why are we setting this to False?

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 7]
```diff
@@ -1610,3 +1610,90 @@ def test_tucker_tensorly(tol_norm_2, tol_max_abs, shape, ranks):
...
+        )
+    )
+    tensors_num = draw(helpers.ints(min_value=1, max_value=10))
+    return dtype, x[0], tensors_num
+
+
+@handle_test(
+    fn_tree="functional.ivy.experimental.batched_outer",
+    data=_batched_outer_data(),
+    test_with_out=st.just(False),
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1671]: Again, this wont be required after you update the implementation as I suggested.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 8]
```diff
@@ -1610,3 +1610,90 @@ def test_tucker_tensorly(tol_norm_2, tol_max_abs, shape, ranks):
...
+def test_batched_outer(*, data, test_flags, backend_fw, fn_name, on_device):
+    input_dtypes, x, tensors_num = data
+    helpers.test_function(
+        backend_to_test=backend_fw,
+        test_flags=test_flags,
+        fn_name=fn_name,
+        on_device=on_device,
+        input_dtypes=input_dtypes,
+        tensors=[x] * tensors_num,
+        test_values=False,
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1682]: Why are we setting this to false?

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 9]
```diff
@@ -1610,3 +1610,90 @@ def test_tucker_tensorly(tol_norm_2, tol_max_abs, shape, ranks):
...
+    )
+
+
+# test adapted from tensorly
+# https://github.com/tensorly/tensorly/blob/main/tensorly/tenalg/tests/test_outer_product.py#L22
+@pytest.mark.skip(
+    reason="ivy.tensordot does not support batched_modes argument for the moment"
+)
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1690]: Can we add a Trello card for adding the `batch_modes` argument to `ivy.tensordot` so that we can keep track of it 😄

**Review by hello-fri-end — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 9 above)
Comment (Madjid-CH) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1690]: there's a todo for super-setting tensordot function [here](https://github.com/unifyai/ivy/issues/21914).
so should I create a trello card anyway ?

Comment (hello-fri-end) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1686]: Oh right, I added it as part of the ToDo list. No need to create a Trello card. Thanks for the heads up :pray: It would be great though if you could pick up that next so that we are able to run this test as well :pray:

**Review by hello-fri-end — CHANGES_REQUESTED** [review]
Thanks for making all the changes @Madjid-CH. I'm really glad with the way this PR is progressing. Could you look into the merge conflicts please - we can merge it afterwards if there aren't any failing test cases on the CI :pray: :smile:

**Comment by Madjid-CH** [comment]
> Thanks for making all the changes @Madjid-CH. I'm really glad with the way this PR is progressing. Could you look into the merge conflicts please - we can merge it afterwards if there aren't any failing test cases on the CI 🙏 😄

certainly

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 9 above)
Comment (Madjid-CH) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py]: I will enable the notifications for tensordot super-setting issue so I'll come back to it when it's implemented.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 10]
```diff
@@ -15,6 +15,24 @@
...
 
+# batched_outer
+@st.composite
+def _batched_outer_data(draw):
+    shape = draw(helpers.get_shape(min_num_dims=2, max_num_dims=3))
+    tensors_num = draw(helpers.ints(min_value=1, max_value=5))
+    dtype, tensors = draw(
+        helpers.dtype_and_values(
+            num_arrays=tensors_num,
+            available_dtypes=["float32", "float64"],
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:26]: Why can't we use `helpers.get_dtypes("float")` here?

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 11]
```diff
@@ -380,6 +398,23 @@ def _get_dtype_value1_value2_cov(
...
 
 
+# higher_order_moment
+@st.composite
+def _higher_order_moment_data(draw):
+    shape = draw(helpers.get_shape(min_num_dims=2, max_num_dims=5))
+    order = draw(helpers.ints(min_value=0, max_value=10))
+    dtype, x = draw(
+        helpers.dtype_and_values(
+            available_dtypes=["float32", "float64"],
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:408]: Same question here.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 12]
```diff
@@ -1034,6 +1110,28 @@ def test_general_inner_product(*, data, test_flags, backend_fw, fn_name, on_devi
...
 
+@handle_test(
+    fn_tree="functional.ivy.experimental.higher_order_moment",
+    data=_higher_order_moment_data(),
+)
+def test_higher_order_moment(*, data, test_flags, backend_fw, fn_name, on_device):
+    input_dtypes, x, order = data
+    if backend_fw == "paddle":
+        # to avoid large dimension results since paddle don't support them
+        order = min(order, 2)
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1121]: I'm not sure if I follow this- could you please explain why this is needed?

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py [hunk 13]
```diff
@@ -763,6 +798,47 @@ def test_adjoint(dtype_x, test_flags, backend_fw, fn_name):
...
 
+@handle_test(
+    fn_tree="functional.ivy.experimental.batched_outer",
+    data=_batched_outer_data(),
+)
+def test_batched_outer(*, data, test_flags, backend_fw, fn_name, on_device):
+    input_dtypes, tensors = data
+    if backend_fw == "paddle":
+        # to avoid large dimension results since paddle don't support them
+        tensors = tensors[:2]
```
Comment (hello-fri-end) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:809]: Don't follow this- could you please explain why this is needed?

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 10 above)
Comment (Madjid-CH) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:26]: because `numpy` backend does not support `bfloat16` so this makes the tests fails.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 12 above)
Comment (Madjid-CH) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1121]: the paddle backend does not support tensors with dimensions bigger than 6 (`len(tensor.shape) > 6`)  and since any addition in  `order `make an additional iteration in `higher_order_moment` function which add more dimensions to the returned tensor.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 13 above)
Comment (Madjid-CH) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:809]: same reason as previously mentioned, more tensors result in bigger dimensions  in the result.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 10 above)
Comment (hello-fri-end) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:26]: Which numpy function is causing the test to fail? Ideally, `bfloat16` should be in the `unsupported_dtypes` for that function.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 12 above)
Comment (hello-fri-end) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py:1121]: Ah, makes sense. Thanks for explaining :+1:

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py (hunk 11 above)
Comment (Madjid-CH) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_linalg.py]: Hey @hello-fri-end , I revised this dtype issue and I added some `with_supported_dtypes `decorator to the functions that made the tests fail except for the paddle backend since it will be fixed in another PR

**Review by hello-fri-end — APPROVED** [review]
Looks like everything has been resolved here and there are no failing tests. Feel free to merge the PR @Madjid-CH . Thanks for the contribution :rocket:
